README.md
# HAProxy Load Balancer
[![Build Status](https://travis-ci.org/million12/docker-haproxy.svg?branch=master)](https://travis-ci.org/million12/docker-haproxy)
[![GitHub Open Issues](https://img.shields.io/github/issues/million12/docker-haproxy.svg)](https://github.com/million12/docker-haproxy/issues)
[![Stars](https://img.shields.io/github/stars/million12/docker-haproxy.svg?style=social&label=Stars)]()
[![Fork](https://img.shields.io/github/forks/million12/docker-haproxy.svg?style=social&label=Fork)]()
[![Release](https://img.shields.io/github/release/million12/docker-haproxy.svg)](http://microbadger.com/images/million12/haproxy.svg)

[![Docker build](http://dockeri.co/image/million12/haproxy)](https://hub.docker.com/r/million12/haproxy/)

Felling like supporting me in my projects use donate button. Thank You!
[![](https://img.shields.io/badge/donate-PayPal-blue.svg)](https://www.paypal.me/POzgo)

HAProxy docker container [million12/haproxy](https://registry.hub.docker.com/u/million12/haproxy/) with ALPN and HTTP/2 support.

### Tags
Please specify tag when deploying for specific version.
Example:

`million12/haproxy:latest`
`million12/haproxy:1.9.2`

# Features

* **Support for HTTP/2** with ALPN
* CentOS 7 based
* Ability to **provide any arguments to haproxy** process
  Any extra parameters provided to `docker run` will be passed directly to `haproxy` command.
  For example, if you run `docker run [run options] million12/haproxy -n 1000` you pass `-n 1000` to haproxy daemon.
* Pretty **lightweight**, only ~100M (with OpenSSL and HAProxy compiled from source).
* **Default [haproxy.cfg](container-files/etc/haproxy/haproxy.cfg) provided** for demonstration purposes. You can easily mount your own or point to different location using `HAPROXY_CONFIG` env.
* **Auto restart when config changes**
  This container comes with inotify to monitor changes in HAProxy config and **reload** HAProxy daemon. The reload is done in a way that no connection is lost.

## ENV variables

|Variable|Default Settings|Notes|
|:--|:--|:--|
|`HAPROXY_CONFIG`|`/etc/haproxy/haproxy.cfg`|If you mount your config to different location, simply edit it.|
|`HAPROXY_PORTS`|`80,443`|Comma separated ports|
|`HAPROXY_ADDITIONAL_CONFIG`|Empty|List of file that inotify should monitor for changes divided by space. Example below. Space separated|
|`HAPROXY_PRE_RESTART_CMD`|Empty|Command to execute before restarting haproxy|
|`HAPROXY_POST_RESTART_CMD`|Empty|Command to execute after successfully restarting haproxy|

## Usage

### Basic

```bash
docker run -ti \
  --cap-add NET_ADMIN \
  -p 80:80 \
  -p 443:443 \
  million12/haproxy
```

### Mount custom config , override some options

```bash
docker run -d \
  --cap-add NET_ADMIN \
  -p 80:80 \
  -v /my-haproxy.cfg:/etc/haproxy/haproxy.cfg \
  million12/haproxy \
  -n 10000
```
Note: in this case config is mounted to its default location, so you don't need to modify
`HAPROXY_CONFIG` variable.

### Monitor additional config files

```bash
docker run -d \
  --cap-add NET_ADMIN \
  -p 80:80 \
  -e HAPROXY_ADDITIONAL_CONFIG='/etc/haproxy/custom1 /etc/haproxy/custom2' \
  million12/haproxy
```

### Check version and build options

`docker run -ti million12/haproxy -vv`

### Stats
The default URL for stats is `http://CONTAINER_IP/admin?stats` with username:password ser to `admin:admin`.

## Authors

Author: Marcin ryzy Ryzycki (<marcin@m12.io>)
Author: Przemyslaw Ozgo (<linux@ozgo.info>)


CHANGELOG
ChangeLog :
===========

2019/01/16 : 1.9.2
    - BUG/MAJOR: cache: fix confusion between zero and uninitialized cache key
    - BUG/MEDIUM: h1: Make sure we destroy an inactive connectin that did shutw.
    - MEDIUM: sessions: Keep track of which connections are idle.
    - BUG/MEDIUM: init: Initialize idle_orphan_conns for first server in server-template
    - BUG/MEDIUM: connection: properly unregister the mux on failed initialization
    - MINOR: checks: Store the proxy in checks.
    - BUG/MEDIUM: checks: Avoid having an associated server for email checks.
    - BUG/MINOR: lua/htx: Respect the reserve when data are send from an HTX applet
    - BUG/MEDIUM: ssl: Disable anti-replay protection and set max data with 0RTT.
    - DOC: Be a bit more explicit about allow-0rtt security implications.
    - BUG/MEDIUM: ssl: missing allocation failure checks loading tls key file
    - BUG/MINOR: backend: don't use url_param_name as a hint for BE_LB_ALGO_PH
    - BUG/MINOR: backend: balance uri specific options were lost across defaults
    - BUG/MINOR: backend: BE_LB_LKUP_CHTREE is a value, not a bit
    - REGTEST: "capture (request|response)" regtest.
    - REGTEST: filters: add compression test
    - REGTESTS: test case for map_regm commit 271022150d
    - REGTESTS: Basic tests for concat,strcmp,word,field,ipmask converters
    - REGTESTS: Basic tests for using maps to redirect requests / select backend
    - DOC: REGTESTS README varnishtest -Dno-htx= define.
    - REGTEST: Switch to vtest.
    - REGTEST: Adapt reg test doc files to vtest.
    - BUILD/MEDIUM: da: Necessary code changes for new buffer API.
    - BUG/MINOR: base64: dec func ignores padding for output size checking
    - MINOR: ssl: add support of aes256 bits ticket keys on file and cli.
    - BUG/MINOR: stick_table: Prevent conn_cur from underflowing
    - MINOR: spoe: Make the SPOE filter compatible with HTX proxies
    - MINOR: h2: add a bit-based frame type representation
    - MEDIUM: mux-h2: remove padlen during headers phase
    - MINOR: mux-h2: remove useless check for empty frame length in h2s_decode_headers()
    - MEDIUM: mux-h2: decode HEADERS frames before allocating the stream
    - MINOR: mux-h2: make h2c_send_rst_stream() use the dummy stream's error code
    - MINOR: mux-h2: add a new dummy stream for the REFUSED_STREAM error code
    - MINOR: mux-h2: fail stream creation more cleanly using RST_STREAM
    - MINOR: buffers: add a new b_move() function
    - MINOR: mux-h2: make h2_peek_frame_hdr() support an offset
    - MEDIUM: mux-h2: handle decoding of CONTINUATION frames
    - BUG/MINOR: mux-h2: set the stream-full flag when leaving h2c_decode_headers()
    - BUG/MINOR: mux-h2: mark end-of-stream after processing response HEADERS, not before
    - BUG/MINOR: mux-h2: only update rxbuf's length for H1 headers
    - MINOR: mux-h2: make h2c_decode_headers() return a status, not a count
    - MINOR: mux-h2: add a new dummy stream : h2_error_stream
    - MEDIUM: mux-h2: make h2c_decode_headers() support recoverable errors
    - BUG/MINOR: mux-h2: detect when the HTX EOM block cannot be added after headers
    - MINOR: mux-h2: check for too many streams only for idle streams
    - MINOR: mux-h2: set H2_SF_HEADERS_RCVD when a HEADERS frame was decoded
    - BUG/MEDIUM: mux-h2: decode trailers in HEADERS frames
    - MINOR: h2: add h2_make_h1_trailers to turn H2 headers to H1 trailers
    - MEDIUM: mux-h2: pass trailers to H1 (legacy mode)
    - MINOR: htx: add a new function to add a block without filling it
    - MINOR: h2: add h2_make_htx_trailers to turn H2 headers to HTX trailers
    - MEDIUM: mux-h2: pass trailers to HTX
    - MINOR: mux-h2: make HTX_BLK_EOM processing idempotent
    - MINOR: h1: make the H1 headers block parser able to parse headers only
    - BUG/MEDIUM: h1: Get the h1m state when restarting the headers parsing
    - MEDIUM: mux-h2: emit HEADERS frames when facing HTX trailers blocks
    - BUG: 51d: Changes to the buffer API in 1.9 were not applied to the 51Degrees code.
    - BUG/MEDIUM: stats: Get the right scope pointer depending on HTX is used or not

2019/01/08 : 1.9.1
    - BUG/MEDIUM: tasks: Decrement tasks_run_queue in tasklet_free().
    - BUG/MEDIUM: log: don't mark log FDs as non-blocking on terminals
    - BUG/MAJOR: stream-int: Update the stream expiration date in stream_int_notify()
    - BUG/MAJOR: connections: Close the connection before freeing it.
    - BUG/MEDIUM: h2: Don't forget to quit the sending_list if SUB_CALL_UNSUBSCRIBE.
    - BUG/MEDIUM: mux-h2: Don't forget to quit the send list on error reports
    - BUG/MEDIUM: mux-h2: don't needlessly wake up the demux on short frames
    - MINOR: mux-h2: only increase the connection window with the first update
    - REGTESTS: remove the expected window updates from H2 handshakes
    - BUG/MINOR: mux-h2: make empty HEADERS frame return a connection error
    - BUG/MEDIUM: mux-h2: mark that we have too many CS once we have more than the max
    - BUG/MEDIUM: mux_h2: Don't add to the idle list if we're full.
    - BUG/MEDIUM: mux-h2: always restart reading if data are available
    - BUG/MINOR: mux-h2: don't check the CS count in h2c_bck_handle_headers()
    - BUG/MEDIUM: dns: Don't prevent reading the last byte of the payload in dns_validate_response()
    - BUG/MEDIUM: dns: overflowed dns name start position causing invalid dns error
    - BUG/MINOR: compression/htx: Don't compress responses with unknown body length
    - BUG/MINOR: compression/htx: Don't add the last block of data if it is empty
    - BUG/MEDIUM: server: Also copy "check-sni" for server templates.
    - BUG/MEDIUM: servers: Don't try to reuse connection if we switched server.
    - BUG/MEDIUM: servers: Fail if we fail to allocate a conn_stream.
    - BUG/MAJOR: servers: Use the list api correctly to avoid crashes.
    - BUG/MAJOR: servers: Correctly use LIST_ELEM().
    - BUG/MAJOR: sessions: Use an unlimited number of servers for the conn list.

container-files/bootstrap.sh
#!/bin/sh

set -u

# User params
HAPROXY_CONFIG=${HAPROXY_CONFIG:="/etc/haproxy/haproxy.cfg"}
HAPROXY_ADDITIONAL_CONFIG=${HAPROXY_ADDITIONAL_CONFIG:=""}
HAPROXY_PORTS=${HAPROXY_PORTS:="80,443"}
HAPROXY_PRE_RESTART_CMD=${HAPROXY_PRE_RESTART_CMD:=""}
HAPROXY_POST_RESTART_CMD=${HAPROXY_POST_RESTART_CMD:=""}
HAPROXY_USER_PARAMS=$@

# Internal params
HAPROXY_PID_FILE="/var/run/haproxy.pid"
HAPROXY_CMD="/usr/local/sbin/haproxy -f ${HAPROXY_CONFIG} ${HAPROXY_USER_PARAMS} -D -p ${HAPROXY_PID_FILE}"
HAPROXY_CHECK_CONFIG_CMD="/usr/local/sbin/haproxy -f ${HAPROXY_CONFIG} -c"

# Iptable commands
LIST_IPTABLES="iptables --list"
ENABLE_SYN_DROP="iptables -I INPUT -p tcp -m multiport --dport $HAPROXY_PORTS --syn -j DROP"
DISABLE_SYN_DROP="iptables -D INPUT -p tcp -m multiport --dport $HAPROXY_PORTS --syn -j DROP"


#######################################
# Echo/log function
# Arguments:
#   String: value to log
#######################################
log() {
  if [[ "$@" ]]; then echo "[`date +'%Y-%m-%d %T'`] $@";
  else echo; fi
}

#######################################
# Dump current $HAPROXY_CONFIG
#######################################
print_config() {
  log "Current HAProxy config $HAPROXY_CONFIG:"
  printf '=%.0s' {1..100} && echo
  cat $HAPROXY_CONFIG
  printf '=%.0s' {1..100} && echo
}

# Check iptables rules modification capabilities
$LIST_IPTABLES > /dev/null 2>&1
# Exit immidiately in case of any errors
if [[ $? != 0 ]]; then
    EXIT_CODE=$?;
    echo "Please enable NET_ADMIN capabilities by passing '--cap-add NET_ADMIN' parameter to docker run command";
    exit $EXIT_CODE;
fi

# Launch HAProxy.
# In the default attached haproxy.cfg `web.server` host is used for back-end nodes.
# If that host doesn't exist in /etc/hosts, create it and point to localhost,
# so HAProxy can start with the default haproxy.cfg config without throwing errors.
grep --silent -e "web.server" /etc/hosts || echo "127.0.0.1 web.server" >> /etc/hosts

log $HAPROXY_CMD && print_config
$HAPROXY_CHECK_CONFIG_CMD
$HAPROXY_CMD
# Exit immidiately in case of any errors or when we have interactive terminal
if [[ $? != 0 ]] || test -t 0; then exit $?; fi
log "HAProxy started with $HAPROXY_CONFIG config, pid $(cat $HAPROXY_PID_FILE)." && log

# Check if config has changed
# Note: also monitor /etc/hosts where the new back-end hosts might be provided.
while inotifywait -q -e create,delete,modify,attrib /etc/hosts $HAPROXY_CONFIG $HAPROXY_ADDITIONAL_CONFIG; do
  if [ -f $HAPROXY_PID_FILE ]; then
    log "Restarting HAProxy due to config changes..." && print_config
    $HAPROXY_CHECK_CONFIG_CMD
    $ENABLE_SYN_DROP
    sleep 0.2
    log "Executing pre-restart hook..."
    $HAPROXY_PRE_RESTART_CMD
    log "Restarting haproxy..."
    $HAPROXY_CMD -sf $(cat $HAPROXY_PID_FILE)
    log "Executing post-restart hook..."
    $HAPROXY_POST_RESTART_CMD
    $DISABLE_SYN_DROP
    log "HAProxy restarted, pid $(cat $HAPROXY_PID_FILE)." && log
  else
    log "Error: no $HAPROXY_PID_FILE present, HAProxy exited."
    break
  fi
done
